File: wxFEFactory/pyapi/layout/controls.cpp
#include <wx/wx.h>
#include "controls.h"

long Text::getAlignStyle()
{
	pyobj style = getStyle(STYLE_TEXTALIGN);
	long wxstyle = 0;
	if (!style.is(None))
	{
		wxcstr align = style.cast<wxString>();
		if (align != wxNoneString) {
			if (align == wxT("center"))
			{
				wxstyle |= wxALIGN_CENTER_HORIZONTAL;
			}
			else if (align == wxT("right"))
			{
				wxstyle |= wxALIGN_RIGHT;
			}
			else if (align == wxT("left"))
			{
				wxstyle |= wxALIGN_LEFT;
			}
		}
	}
	return wxstyle;
}

void RadioBox::applyStyle()
{
	View::applyStyle();

	pyobj style;

	style = getStyle(STYLE_FLEXDIRECTION);
	if (!style.is(None))
	{
		wxcstr dir = style.cast<wxString>();
		if (dir != wxNoneString) {
			long style = ctrl().GetWindowStyle();
			if (dir == wxT("row"))
			{
				style |= wxRA_SPECIFY_ROWS;
			}
			else if (dir == wxT("column"))
			{
				style |= wxRA_SPECIFY_COLS;
			}
			ctrl().SetWindowStyle(style);
		}
	}
}

void init_controls(py::module & m)
{
	using namespace py::literals;

	auto className = "className"_a = None;
	auto style = "style"_a = None;
	auto label = "label"_a;
	auto type = "type"_a = wxEmptyString;
	auto wxstyle = "wxstyle"_a = 0;
	auto evt_fn = "fn"_a;
	auto evt_reset = "reset"_a = true;

	py::class_t<Button, Control>(m, "Button")
		.def(py::init<wxcstr, pyobj, pyobj, pyobj>(),
			label, "onclick"_a = None, className, style)
		.def("setOnClick", &Button::setOnClick, "onclick"_a, evt_reset)
		.def("getLabel", &Button::getLabel)
		.def("setLabel", &Button::setLabel)
		.def("click", &Button::click)
		.def_property("label", &Button::getLabel, &Button::setLabel);

	py::class_t<ToggleButton, Control>(m, "ToggleButton")
		.def(py::init<wxcstr, bool, pyobj, pyobj, pyobj>(),
			label, "checked"_a = false, "onchange"_a = None, className, style)
		.def("getLabel", &ToggleButton::getLabel)
		.def("setLabel", &ToggleButton::setLabel)
		.def("trigger", &ToggleButton::trigger)
		.def("setOnChange", &ToggleButton::setOnChange, "onchange"_a, evt_reset)
		.def_property("label", &ToggleButton::getLabel, &ToggleButton::setLabel)
		.def_property("checked", &ToggleButton::getChecked, &ToggleButton::setChecked);

	py::class_t<CheckBox, Control>(m, "CheckBox")
		.def(py::init<wxcstr, bool, bool, pyobj, pyobj, pyobj>(),
			label, "checked"_a = false, "alignRight"_a = false, "onchange"_a = None, className, style)
		.def("getLabel", &CheckBox::getLabel)
		.def("setLabel", &CheckBox::setLabel)
		.def("trigger", &CheckBox::trigger)
		.def_property("label", &CheckBox::getLabel, &CheckBox::setLabel)
		.def_property("checked", &CheckBox::getChecked, &CheckBox::setChecked);

	py::class_t<Text, Control>(m, "Text")
		.def(py::init<wxcstr, pyobj, pyobj>(), label, className, style)
		.def("getLabel", &Text::getLabel)
		.def("setLabel", &Text::setLabel)
		.def_property("label", &Text::getLabel, &Text::setLabel);

	py::class_t<TextInput, Control>(m, "TextInput")
		.def(py::init<wxcstr, wxcstr, bool, bool, long, pyobj, pyobj>(),
			"value"_a = wxEmptyString, type, "readonly"_a = false, "multiline"_a = false, wxstyle, className, style)
		.def("setOnEnter", &TextInput::setOnEnter, evt_fn, evt_reset)
		.def("setOnChar", &TextInput::setOnChar, evt_fn, evt_reset)
		.def("appendText", &TextInput::appendText)
		.def("writeText", &TextInput::writeText)
		.def("selectAll", &TextInput::selectAll)
		.def("clear", &TextInput::clear)
		.def_property("value", &TextInput::getValue, &TextInput::setValue)
		.def_property("selection", &TextInput::getSelection, &TextInput::setSelection);

	py::class_t<SearchCtrl, Control>(m, "SearchCtrl")
		.def(py::init<wxcstr, bool, bool, long, pyobj, pyobj>(),
			"value"_a = wxEmptyString, "search_button"_a = true, "cancel_button"_a = true, wxstyle, className, style)
		.def("setOnSubmit", &SearchCtrl::setOnSubmit, evt_fn, evt_reset)
		.def("setOnCancel", &SearchCtrl::setOnCancel, evt_fn, evt_reset)
		.def_property("value", &SearchCtrl::getValue, &SearchCtrl::setValue);

	py::class_t<SpinCtrl, Control>(m, "SpinCtrl")
		.def(py::init<wxcstr, int, int, int, pyobj, pyobj>(),
			"value"_a = wxEmptyString, "min"_a=0, "max"_a=100, "initial"_a=0, className, style)
		.def_property("value", &SpinCtrl::getValue, &SpinCtrl::setValue)
		.def_property("min", &SpinCtrl::getMin, &SpinCtrl::setMin)
		.def_property("max", &SpinCtrl::getMax, &SpinCtrl::setMax);

	py::class_t<ColorPicker, Control>(m, "ColorPicker")
		.def(py::init<uint, pyobj, pyobj, pyobj>(),
			"color"_a = 0, "onchange"_a = None, className, style)
		.def_property("color", &ColorPicker::getColor, &ColorPicker::setColor);

	auto choices = "choices"_a = None,
		onselect = "onselect"_a = None;

	py::class_<BaseControlWithItems, Control>(m, "BaseControlWithItems")
		.def("getText", &BaseControlWithItems::getText, "pos"_a = -1)
		.def("getTexts", &BaseControlWithItems::getTexts)
		.def("setText", &BaseControlWithItems::setText, "text"_a, "pos"_a = -1)
		.def("getSelection", &BaseControlWithItems::getSelection)
		.def("setSelection", &BaseControlWithItems::setSelection, "pos"_a, "handle"_a = false)
		.def("getCount", &BaseControlWithItems::getCount)
		.def("__getitem__", &BaseControlWithItems::__getitem__)
		.def("__setitem__", &BaseControlWithItems::__setitem__)
		.def("__len__", &BaseControlWithItems::getCount)
		.def("prev", &BaseControlWithItems::prev, "handle"_a = true)
		.def("next", &BaseControlWithItems::next, "handle"_a = true)
		.def_property("text", &BaseControlWithItems::getText1, &BaseControlWithItems::setText1)
		.def_property("index", &BaseControlWithItems::getSelection, &BaseControlWithItems::doSetSelection)
		.def_property_readonly("count", &BaseControlWithItems::getCount);

	py::class_<ControlWithItems, BaseControlWithItems>(m, "ControlWithItems")
		.def("append", &ControlWithItems::append, "text"_a)
		.def("insert", &ControlWithItems::insert, "text"_a, "pos"_a)
		.def("appendItems", &ControlWithItems::appendItems, "choices"_a)
		.def("insertItems", &ControlWithItems::insertItems, "choices"_a, "pos"_a)
		.def("setItems", &ControlWithItems::setItems, "choices"_a)
		.def("__delitem__", &ControlWithItems::pop)
		.def("pop", &ControlWithItems::pop, "pos"_a)
		.def("clear", &ControlWithItems::clear);

	py::class_t<ListBox, ControlWithItems>(m, "ListBox")
		.def(py::init<pyobj, pyobj, pyobj, pyobj>(),
			choices, onselect, className, style);

	py::class_t<CheckListBox, ListBox>(m, "CheckListBox")
		.def(py::init<pyobj, pyobj, pyobj, pyobj>(),
			choices, onselect, className, style)
		.def("getCheckedItems", &CheckListBox::getCheckedItems)
		.def("setCheckedItems", &CheckListBox::setCheckedItems)
		.def("checkAll", &CheckListBox::checkAll, "checked"_a = true)
		.def("reverseCheck", &CheckListBox::reverseCheck);

	py::class_t<RearrangeList, CheckListBox>(m, "RearrangeList")
		.def(py::init<pyobj, pyobj, pyobj, pyobj, pyobj>(),
			choices, "order"_a = py::list(), onselect, className, style)
		.def("moveUp", &RearrangeList::moveUp)
		.def("moveDown", &RearrangeList::moveDown);

	py::class_t<Choice, ControlWithItems>(m, "Choice")
		.def(py::init<pyobj, pyobj, pyobj, pyobj>(),
			choices, onselect, className, style);

	py::class_t<ComboBox, ControlWithItems>(m, "ComboBox")
		.def(py::init<wxcstr, pyobj, pyobj, pyobj, pyobj>(),
			type, choices, onselect, className, style)
		.def("setOnEnter", &ComboBox::setOnEnter, evt_fn, evt_reset)
		.def("autoComplete", &ComboBox::autoComplete)
		.def_property("value", &ComboBox::getValue, &ComboBox::setValue);

	py::class_t<RadioBox, BaseControlWithItems>(m, "RadioBox")
		.def(py::init<wxcstr, pyobj, pyobj, pyobj, pyobj>(),
			label, choices, onselect, className, style);

	py::class_t<Hr, Control>(m, "Hr")
		.def(py::init<bool, pyobj, pyobj>(),
			"vertical"_a = false, className, style);

	auto pyFilePickerCtrl = py::class_t<FilePickerCtrl, Control>(m, "FilePickerCtrl")
		.def(py::init<wxcstr, wxcstr, wxcstr, long, pyobj, pyobj>(),
			"path"_a = wxEmptyString, "msg"_a = wxEmptyString, "wildcard"_a = (const char*)wxFileSelectorDefaultWildcardStr,
			"wxstyle"_a = (long)(wxFLP_DEFAULT_STYLE | wxFLP_SMALL), className, style)
		.def("setOnChange", &FilePickerCtrl::setOnChange, evt_fn, evt_reset)
		.def_property("path", &FilePickerCtrl::getPath, &FilePickerCtrl::setPath)
		.ptr();

#define ATTR_FLP_STYLE(name) ATTR_INT(pyFilePickerCtrl, name, wxFLP_)
	ATTR_FLP_STYLE(FILE_MUST_EXIST),
		ATTR_FLP_STYLE(OPEN),
		ATTR_FLP_STYLE(OVERWRITE_PROMPT),
		ATTR_FLP_STYLE(SAVE),
		ATTR_FLP_STYLE(SMALL),
		ATTR_FLP_STYLE(USE_TEXTCTRL);
#undef ATTR_FLP_STYLE

	auto pyDirPickerCtrl = py::class_t<DirPickerCtrl, Control>(m, "DirPickerCtrl")
		.def(py::init<wxcstr, wxcstr, long, pyobj, pyobj>(),
			"path"_a = wxEmptyString, "msg"_a = wxEmptyString, "wxstyle"_a = (long)(wxDIRP_DEFAULT_STYLE | wxDIRP_SMALL), className, style)
		.def("setOnChange", &DirPickerCtrl::setOnChange, evt_fn, evt_reset)
		.def_property("path", &DirPickerCtrl::getPath, &DirPickerCtrl::setPath)
		.ptr();

#define ATTR_DIRP_STYLE(name) ATTR_INT(pyDirPickerCtrl, name, wxDIRP_)
	ATTR_DIRP_STYLE(DIR_MUST_EXIST),
		ATTR_DIRP_STYLE(SMALL),
		ATTR_DIRP_STYLE(USE_TEXTCTRL);
#undef ATTR_FLP_STYLE
}

File: wxFEFactory/python/lib/ui/controls.py
from lib.win32.keys import WXK
from .view import View, Control, EventFunctor, event_binder, value_property
from . import wx


class Button(Control):
    """普通按钮"""
    wxtype = wx.Button

    def __init__(self, label, onclick=None, **kwargs):
        super().__init__(wxparams={'label': label}, **kwargs)
        if onclick:
            self.set_onclick(onclick)

    set_onclick = event_binder(wx.EVT_BUTTON)

    def click(self):
        self.post_event(wx.CommandEvent)


class BitmapButton(Button):
    """图像按钮"""
    def __init__(self, src, **kwargs):
        self.src = src
        super().__init__(**kwargs)

    def render(self, parent):
        bitmap = wx.Bitmap(self.src)
        self.bind_wx(wx.BitmapButton(parent, label=bitmap, **self.wxparams))
        del self.src, self.onclick


class CheckableControl(Control):
    def __init__(self, checked=False, onchange=None, **kwargs):
        super().__init__(**kwargs)
        if checked:
            self.SetValue(True)
        if onchange:
            self.set_onchange(onchange)

    def set_onchange(self, fn, reset=True):
        self.bind_event(self.wxevent, fn, reset)

    def toggle(self, value=None, post=True):
        """切换状态"""
        if value is None:
            value = not self.GetValue()
        self.SetValue(value)
        if post:
            self.post_event(self.wxevent)
        else:
            self.call_event(self.wxevent)

    checked = value = value_property


class ToggleButton(CheckableControl):
    """开关按钮"""
    wxtype = wx.ToggleButton
    wxevent = wx.EVT_TOGGLEBUTTON

    def __init__(self, label, **kwargs):
        super().__init__(wxparams={'label': label}, **kwargs)


class CheckBox(CheckableControl):
    """复选框"""
    wxtype = wx.CheckBox
    wxevent = wx.EVT_CHECKBOX

    def __init__(self, label="", align_right=False, **kwargs):
        if align_right:
            kwargs['wxstyle'] = kwargs.get('wxstyle', 0) | wx.ALIGN_RIGHT
        kwargs['wxparams'] = {'label': label}
        super().__init__(**kwargs)


class StaticBitmap(Control):
    """静态图片"""
    wxtype = wx.StaticBitmap

    def __init__(self, label, **kwargs):
        kwargs['wxparams'] = {'label': label}
        super().__init__(**kwargs)

    def render(self, parent):
        bitmap = wx.Bitmap(self.label)
        self.bind_wx(wx.StaticBitmap(parent.wxwindow, label=bitmap, **self.wxparams))
        del self.label


class Text(Control):
    """静态文本"""
    wxtype = wx.StaticText

    def __init__(self, label, **kwargs):
        super().__init__(wxparams={'label': label}, **kwargs)

    def apply_style(self):
        super().apply_style()
        align = self.computed_style.get('align', None)
        if align is not None:
            self.wxstyle |= align


class TextInput(Control):
    """文本输入框"""
    wxtype = wx.TextCtrl

    def __init__(self, value="", multiline=False, readonly=False, type=None, **kwargs):
        wxstyle = kwargs.get('wxstyle', 0)
        if multiline:
            wxstyle |= wx.TE_MULTILINE
        if readonly:
            wxstyle |= wx.TE_READONLY
        if type == "password":
            wxstyle |= wx.TE_PASSWORD
        if wxstyle:
            kwargs['wxstyle'] = wxstyle
        kwargs['wxparams'] = {'value': value}
        super().__init__(**kwargs)

    set_onenter = event_binder(wx.EVT_TEXT_ENTER)
    set_onchar = event_binder(wx.EVT_CHAR)

    @property
    def selection(self):
        return self.GetSelection()

    @selection.setter
    def selection(self, value):
        self.SetSelection(value)

    value = value_property


class SearchCtrl(Control):
    """搜索框"""
    wxtype = wx.SearchCtrl

    def __init__(self, value="", search_button=True, cancel_button=False, **kwargs):
        super().__init__(wxparams={'value': value}, **kwargs)
        if not search_button:
            self.ShowSearchButton(False)
        if cancel_button:
            self.ShowCancelButton(True)

    set_onsubmit = event_binder(wx.EVT_SEARCH)
    set_oncancel = event_binder(wx.EVT_SEARCH_CANCEL)
    value = value_property


class SpinCtrl(Control):
    """数字输入框"""
    wxtype = wx.SpinCtrl

    def __init__(self, value="", min=0, max=100, initial=0, wxstyle=wx.SP_ARROW_KEYS | wx.ALIGN_RIGHT, **kwargs):
        kwargs['wxparams'] = dict(value=value, min=min, max=max, initial=initial)
        kwargs['wxstyle'] = wxstyle
        super().__init__(**kwargs)

    set_onchange = event_binder(wx.EVT_SPINCTRL)
    set_onenter = event_binder(wx.EVT_TEXT_ENTER)


class ColorPicker(Control):
    """颜色选择器"""
    wxtype = wx.ColourPickerCtrl

    def __init__(self, color=None, onchange=None, **kwargs):
        super().__init__(wxparams={'color': color}, **kwargs)
        self.set_onchange(onchange)

    set_onchange = event_binder(wx.EVT_COLOURPICKER_CHANGED)


class ItemContainer(Control):
    @property
    def count(self):
        return self.GetCount()

    @property
    def index(self):
        return self.GetSelection()

    @index.setter
    def index(self, n):
        self.SetSelection(n)

    @property
    def text(self):
        return self.GetString(self.GetSelection())


class ControlWithItems(ItemContainer):
    def __init__(self, onselect=None, **kwargs):
        super().__init__(**kwargs)
        self.set_onselect(onselect)

    def set_onselect(self, fn, reset=True):
        self.bind_event(self.wxevent, fn, reset)

    def post_select(self):
        self.post_event(self.wxevent)

    def set_selection(self, n, trigger=False):
        self.SetSelection(n)
        if trigger:
            self.post_select()
        return self


class ListBox(ControlWithItems):
    """列表框"""
    wxtype = wx.ListBox
    wxevent = wx.EVT_LISTBOX

    def __init__(self, choices=None, **kwargs):
        if choices is not None:
            kwargs['wxparams'] = {'choices': choices}
        super().__init__(**kwargs)


class CheckListBox(ListBox):
    """带复选列表框"""
    wxtype = wx.CheckListBox


class RearrangeList(CheckListBox):
    """可排序复选列表框"""
    wxtype = wx.RearrangeList

    def __init__(self, order=None, **kwargs):
        super().__init__(wxparams={'order': order}, **kwargs)


class Choice(ControlWithItems):
    """简易下拉框"""
    wxtype = wx.Choice
    wxevent = wx.EVT_CHOICE

    def __init__(self, choices=None, **kwargs):
        if choices is not None:
            kwargs['wxparams'] = {'choices': choices}
        super().__init__(**kwargs)


class ComboBox(ControlWithItems):
    """下拉框"""
    wxtype = wx.ComboBox
    wxevent = wx.EVT_COMBOBOX

    def __init__(self, choices=None, value="", **kwargs):
        wxparams = {'value': value}
        if choices is not None:
            wxparams['choices'] = choices
        kwargs['wxparams'] = wxparams
        super().__init__(**kwargs)

    set_onenter = event_binder(wx.EVT_TEXT_ENTER)
    value = value_property


class RadioBox(ControlWithItems):
    """选择框"""
    wxtype = wx.RadioBox
    wxevent = wx.EVT_RADIOBOX

    def __init__(self, title="", choices=None, **kwargs):
        wxparams = {'title': title}
        if choices:
            wxparams['choices'] = choices
        kwargs['wxparams'] = wxparams
        super().__init__(**kwargs)

    def apply_style(self):
        super().apply_style()
        direction = self.computed_style.get('flex-direction', None)
        if direction is not None:
            if direction == 'row':
                self.wxstyle |= wx.RA_SPECIFY_ROWS
            elif direction == 'column':
                self.wxstyle |= wx.RA_SPECIFY_COLS


class Hr(Control):
    """静态线"""
    wxtype = wx.StaticLine

    def __init__(self, vertical=False, **kwargs):
        kwargs['wxstyle'] = wx.VERTICAL if vertical else wx.HORIZONTAL
        super().__init__(**kwargs)


class FilePickerCtrl(Control):
    """文件选择器"""
    wxtype = wx.FilePickerCtrl

    def __init__(self, path="", message="", wildcard="", **kwargs):
        kwargs['wxparams'] = dict(path=path, message=message, wildcard=wildcard)
        super().__init__(**kwargs)

    set_onchange = event_binder(wx.EVT_FILEPICKER_CHANGED)


class DirPickerCtrl(Control):
    """目录选择器"""
    wxtype = wx.DirPickerCtrl

    def __init__(self, path="", message="", **kwargs):
        kwargs['wxparams'] = dict(path=path, message=message)
        super().__init__(**kwargs)

    set_onchange = event_binder(wx.EVT_DIRPICKER_CHANGED)


class TreeCtrl(Control):
    """树控件"""
    wxtype = wx.TreeCtrl

    def __init__(self, wxstyle=wx.TR_HAS_BUTTONS | wx.TR_SINGLE, **kwargs):
        super().__init__(wxstyle=wxstyle, **kwargs)

    set_on_item_activated = event_binder(wx.EVT_TREE_ITEM_ACTIVATED, pass_event=True)


class StatusBar(Control):
    wxtype = wx.StatusBar

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.GetParent().SetStatusBar(self.wxwindow)

    def render(self, parent):
        self.wxparams.pop('pos')
        self.wxparams.pop('size')
        super().render(parent)


# exui


def Label(text):
    return Text(text, class_="input_label expand")


class HotkeyCtrl(TextInput):
    def __init__(self, *args, **kwargs):
        self.value = None
        self.code = None
        self.mode = None
        kwargs['wxstyle'] = wx.TE_PROCESS_ENTER
        super().__init__(*args, **kwargs)
        self.set_on_keydown(self.onkey)

    def onkey(self, _, event):
        code = event.GetKeyCode()
        if WXK.ismod(code):
            return
        mod = event.GetModifiers()
        self.handle_key(code, mod)
        return True

    def handle_key(self, code, mod):
        self.value = WXK.getname(code, mod)
        self.code = code
        self.mode = mod
